contracts/wrapped-polyx/lib.rs

```diff
@@ -0,0 +1,456 @@
+        #[ink(constructor)]
+        pub fn new(ticker: Ticker) -> Self {
+          ink_lang::utils::initialize_contract(|contract| {
+              Self::new_init(contract, ticker)
+          })
+        }
+
+        fn new_init(&mut self, ticker: Ticker) {
+          // The contract should always have an identity.
+          self.did = self.get_did(Self::env().account_id()).unwrap();
+          self.ticker = ticker;
+          self.initialized = false;
+        }
+
+        fn create_wrapped_polyx(&mut self) -> Result<()> {
+            let api = Api::new();
+            // Create asset.
+            api.call().asset().create_asset(
+              AssetName(b"Wrapped POLYX".to_vec()),
+              self.ticker,
+              true, // Divisible token.
+              //TODO: Create Other asset type for wrapped tokens
+              AssetType::EquityCommon,
+              vec![],
+              None,
+              true // Disable Investor uniqueness requirements.
+            ).submit()?;
+            // Pause compliance rules to allow transfers.
+            api.call().compliance_manager().pause_asset_compliance(self.ticker).submit()?;
+            Ok(())
+        }
+
+        fn get_did(&self, acc: AccountId) -> Result<IdentityId> {
+            Self::env().extension().get_key_did(acc)?
+              .map(|did| did.into())
+              .ok_or(Error::MissingIdentity)
+        }
+
+        fn get_caller_did(&self) -> Result<IdentityId> {
+          self.get_did(Self::env().caller())
+        }
+
+        fn ensure_has_portfolio(&self, did: IdentityId) -> Result<PortfolioId> {
+          self.portfolios.get(did).ok_or(Error::NoPortfolio)
+        }
+
+        fn ensure_no_portfolio(&self, did: IdentityId) -> Result<()> {
+            if self.portfolios.get(did).is_some() {
+              return Err(Error::AlreadyHavePortfolio);
+            }
+            Ok(())
+        }
+
+        fn ensure_initialized(&self) -> Result<()> {
+            if !self.initialized {
+              return Err(Error::NotInitialized);
+            }
+            Ok(())
+        }
+
+        fn init_asset_and_venue(&mut self) -> Result<()> {
+            if self.initialized {
+              return Err(Error::AlreadyInitialized);
+            }
+            // Create tickers.
+            self.create_wrapped_polyx()?;
+
+            let api = Api::new();
+            // Get the next venue id.
+            let id = api.query().settlement().venue_counter()
+              .map(|v| v.into())?;
+            // Create Venue.
+            api.call().settlement().create_venue(
+              VenueDetails(b"Contract Venue".to_vec()),
+              vec![],
+              VenueType::Other
+            ).submit()?;
+            // Save venue id.
+            self.venue = id;
+            self.initialized = true;
+            Ok(())
+        }
+
+        #[ink(message, payable)]
+        pub fn init(&mut self) -> Result<()> {
+            self.init_asset_and_venue()
+        }
```

Comment:
For contracts that need to create an asset, we might want to pass the `ticker` as a parameter to the `init` call.  This will prevent someone from watching for contract deploys and front-running the ticker registration.